### PR TITLE
Feature/139 add user deploy support

### LINKED ISSUE #139: [FEATURE] add user deploy support
enhancement over user retrieve which was handled in #64.

image

- [x] create / update user with basic data (key, login, name, email, notificationEmail, api-user, locked, active, mustChangePassword, DefaultBusinessUnit)
- [x] set SSO & FederationID
- [x] add global user role
- [x] remove global user role
- add BU specific user role? - **deemed not possible**
- remove BU specific user role? - **deemed not possible**
- [x] add associated BU
- [x] remove associated BU
-  Make user permissions readable (currently IDs only) --> #772 
-  add user permissions --> #772 
-  remove user permissions --> #772

### PR DESCRIPTION
**Pull request by JoernBerkefeld**
# PR details

## Dependencies

- :warning: awaits merge & release of PR https://github.com/DougMidgley/SFMC-SDK/pull/191 :warning:
- --> done with PR #788

## What changes did you make? (Give an overview)

- closes #139 https://github.com/DougMidgley/SFMC-SDK/pull/191

image

## to dos

- [x] create / update user with basic data (key, login, name, email, notificationEmail, api-user, locked, active, mustChangePassword, DefaultBusinessUnit)
- [x] set SSO & FederationID
- [x] add global user role
- [x] remove global user role
-  add/ BU specific user role? --> #806
- [x] add associated BU
- [x] remove associated BU
-  Make user permissions readable (currently IDs only) --> #772 
-  add user permissions --> #772 
-  remove user permissions --> #772 

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] test scripts updated
- [x] Wiki updated (via #722)

### CODE REVIEW

**Comment by JoernBerkefeld** [comment]
support for removing roles added.

File: lib/metadataTypes/Role.js [hunk 1]
```diff
@@ -36,29 +36,37 @@ class Role extends MetadataType {
...
-                operator: 'equals',
-                rightOperand: false,
-            },
-        };
-        if (key) {
-            // move original filter down one level into rightOperand and add key filter into leftOperand
+        let requestParams = null;
+        if (retrieveDir) {
+            // not needed during caching but certainly during retrieve
+            fields.push('PermissionSets');
```
Comment (DougMidgley) [inline lib/metadataTypes/Role.js:42]: is there any impact of this? previously roles were retrieving fine?

File: lib/metadataTypes/definitions/Role.definition.js [hunk 2]
```diff
@@ -76,7 +76,7 @@ module.exports = {
             template: false,
         },
         PermissionSets: {
-            retrieving: true,
+            retrieving: false, // added in retrieve method if not run for caching
```
Comment (DougMidgley) [inline lib/metadataTypes/definitions/Role.definition.js:79]: so reducing scope where not needed basically?

Comment (JoernBerkefeld) [reply lib/metadataTypes/definitions/Role.definition.js:79]: No, it was indeed. we have a central helper method in our generic soap getter which retrieves the list of fields that are relevant for the retrieve request.
We COULD upgrade all definition files with a new attribute "caching" on each field or we use the workaround that im using here

File: lib/metadataTypes/User.js [hunk 3]
```diff
@@ -0,0 +1,930 @@
...
+                                .BusinessUnitId;
+                        userBUresults[config.ID] = userBUresults[config.ID] || {
+                            add: [],
+                            delete: [],
+                        };
+                        userBUresults[config.ID][
+                            config.BusinessUnitAssignmentConfiguration.IsDelete ? 'delete' : 'add'
+                        ] = Array.isArray(buArr) ? buArr : [buArr];
+                    } else {
+                        // TODO: handle (so far unknown) error
```
Comment (DougMidgley) [inline lib/metadataTypes/User.js:316]: can we at least log something here - otherwise will be one of those needle in haystack thigns

**Review by DougMidgley — CHANGES_REQUESTED** [review]

File: lib/metadataTypes/Role.js (hunk 1 above)
Comment (JoernBerkefeld) [reply lib/metadataTypes/Role.js]: done

File: lib/metadataTypes/User.js (hunk 3 above)
Comment (JoernBerkefeld) [reply lib/metadataTypes/User.js:316]: done

File: lib/metadataTypes/User.js [hunk 4]
```diff
@@ -0,0 +1,930 @@
...
+                        );
+                    }
+                    if (buResult['delete']?.length) {
+                        this.userIdBuMap[userId] = this.userIdBuMap[userId].filter(
+                            (item) => !buResult['delete'].includes(item)
+                        );
+                    }
+                }
+            }
+            // console.log('buResponse', JSON.stringify(buResponse));
```
[the remark this answers is not in the document]
Comment (JoernBerkefeld) [reply lib/metadataTypes/User.js:348]: done

**Pull request merged**